File: docs/js/genetic-function.js
// AG - minimizacao de funcao com codificacao binaria (Atividade 13.1)

var GF_NBITS = 4;
var GF_INIT_POP = ['1001', '1000', '0010', '0001'];
var GF_PCROSS = 1.0;

function gf_r2(n) { return Math.floor(n * 100 + 0.5) / 100; }
function gf_r4(n) { return Math.floor(n * 1e4 + 0.5) / 1e4; }

function gf_decode(ch) { return parseInt(ch, 2); }
function gf_f(x) { return gf_r2(Math.pow(x - 4, 2) - Math.pow(x - 8, 3) + 5); }

function GF_LCG(seed) { this.state = (seed & 0x7FFFFFFF) || 1; }
GF_LCG.prototype.next = function () {
  this.state = (this.state * 16807) % 2147483647;
  return this.state / 2147483647;
};

function gf_evalPop(pop) {
  return pop.map(function (ch) {
    var x = gf_decode(ch);
    return { chrom: ch, x: x, f: gf_f(x) };
  });
}

function gf_fitness(rows) {
  var fmax = Math.max.apply(null, rows.map(function (r) { return r.f; }));
  rows.forEach(function (r) { r.fit = gf_r4(fmax - r.f + 1.0); });
  var total = rows.reduce(function (a, r) { return a + r.fit; }, 0);
  var acc = 0;
  rows.forEach(function (r) {
    r.prob = total > 0 ? gf_r4(r.fit / total) : gf_r4(1.0 / rows.length);
    acc += r.prob; r.cum = gf_r4(acc);
  });
}

function gf_roulette(rows, rng) {
  var r = rng.next();
  for (var i = 0; i < rows.length; i++) if (r <= rows[i].cum) return i;
  return rows.length - 1;
}

function gf_crossover(p1, p2, point) {
  return [p1.slice(0, point) + p2.slice(point), p2.slice(0, point) + p1.slice(point)];
}

function gf_mutate(ch, bit) {
  var a = ch.split('');
  a[bit] = a[bit] === '1' ? '0' : '1';
  return a.join('');
}

function gf_run(seed, n_gen, p_mut) {
  seed = seed === undefined ? 13 : seed;
  n_gen = n_gen === undefined ? 5 : n_gen;
  p_mut = p_mut === undefined ? 0.1 : p_mut;

  var rngSel = new GF_LCG(seed);
  var rngCx = new GF_LCG(seed * 7 + 3);
  var rngMut = new GF_LCG(seed * 13 + 5);

  var pop = GF_INIT_POP.slice();
  var best = null;
  var generations = [];

  for (var g = 1; g <= n_gen; g++) {
    var rows = gf_evalPop(pop);
    gf_fitness(rows);

    var genBest = rows.reduce(function (a, b) { return b.f < a.f ? b : a; });
    if (best === null || genBest.f < best.f)
      best = { chrom: genBest.chrom, x: genBest.x, f: genBest.f, gen: g };

    var pairs = [], children = [], crossOps = [], mutOps = [];
    var nPairs = Math.floor(pop.length / 2);
    for (var k = 0; k < nPairs; k++) {
      var i1 = gf_roulette(rows, rngSel);
      var i2 = gf_roulette(rows, rngSel);
      var p1 = pop[i1], p2 = pop[i2];
      pairs.push({ p1: p1, i1: i1, p2: p2, i2: i2 });
      var c1, c2, point = null;
      if (rngCx.next() <= GF_PCROSS) {
        point = 1 + Math.floor(rngCx.next() * (GF_NBITS - 1));
        var cc = gf_crossover(p1, p2, point); c1 = cc[0]; c2 = cc[1];
      } else { c1 = p1; c2 = p2; }
      crossOps.push({ p1: p1, p2: p2, point: point, c1: c1, c2: c2 });
      children.push(c1, c2);
    }

    var mutated = [];
    children.forEach(function (ch) {
      var cur = ch;
      for (var b = 0; b < GF_NBITS; b++) {
        if (rngMut.next() <= p_mut) {
          var nw = gf_mutate(cur, b);
          mutOps.push({ before: cur, bit: b, after: nw });
          cur = nw;
        }
      }
      mutated.push(cur);
    });

    var newPop = mutated.slice();
    var elite = genBest.chrom;
    var newRows = gf_evalPop(newPop);
    var worstIdx = 0;
    for (var i = 1; i < newRows.length; i++) if (newRows[i].f > newRows[worstIdx].f) worstIdx = i;
    var replaced = newPop[worstIdx];
    newPop[worstIdx] = elite;

    generations.push({
      gen: g, pop: pop.slice(), rows: rows,
      gen_best: { chrom: genBest.chrom, x: genBest.x, f: genBest.f },
      pairs: pairs, cross: crossOps, children: children.slice(),
      mutations: mutOps, mutated: mutated.slice(),
      elite: elite, replaced: replaced, new_pop: newPop.slice(),
      best_so_far: { chrom: best.chrom, x: best.x, f: best.f, gen: best.gen },
    });
    pop = newPop;
  }

  return {
    config: {
      expr: 'f(x) = (x-4)^2 - (x-8)^3 + 5', domain: [0, 15],
      n_bits: GF_NBITS, n_gen: n_gen, p_mut: p_mut, p_cross: GF_PCROSS,
      seed: seed, objective: 'minimizar',
    },
    init_pop: GF_INIT_POP.slice(),
    generations: generations,
    final_pop: gf_evalPop(pop),
    best: best,
  };
}

File: algorithms/genetic_function.py
from __future__ import annotations
import math

# f(x) = (x-4)^2 - (x-8)^3 + 5, x inteiro em [0,15], codificacao binaria de 4 bits
N_BITS   = 4
X_MIN    = 0
X_MAX    = 15
INIT_POP = ['1001', '1000', '0010', '0001']

# probabilidade de cruzamento e numero de geracoes
P_CROSS  = 1.0
N_GEN    = 5


def _r2(x: float) -> float:
    return math.floor(x * 100 + 0.5) / 100


def _r4(x: float) -> float:
    return math.floor(x * 1e4 + 0.5) / 1e4


def _decode(chrom: str) -> int:
    return int(chrom, 2)


def _f(x: int) -> float:
    return _r2((x - 4) ** 2 - (x - 8) ** 3 + 5)


class _LCG:
    def __init__(self, seed: int):
        self.state = seed & 0x7FFFFFFF

    def next(self) -> float:
        self.state = (self.state * 16807) % 2147483647
        return self.state / 2147483647.0


def _eval_pop(pop: list[str]) -> list[dict]:
    rows = []
    for ch in pop:
        x = _decode(ch)
        rows.append({'chrom': ch, 'x': x, 'f': _f(x)})
    return rows


def _fitness(rows: list[dict]) -> None:
    # minimizacao via roleta: fitness = (f_pior - f) + 1 para manter positivo
    fmax = max(r['f'] for r in rows)
    for r in rows:
        r['fit'] = _r4(fmax - r['f'] + 1.0)
    total = sum(r['fit'] for r in rows)
    acc = 0.0
    for r in rows:
        r['prob'] = _r4(r['fit'] / total) if total > 0 else _r4(1.0 / len(rows))
        acc += r['prob']
        r['cum'] = _r4(acc)


def _roulette(rows: list[dict], rng: _LCG) -> int:
    r = rng.next()
    for i, row in enumerate(rows):
        if r <= row['cum']:
            return i
    return len(rows) - 1


def _crossover(p1: str, p2: str, point: int) -> tuple[str, str]:
    c1 = p1[:point] + p2[point:]
    c2 = p2[:point] + p1[point:]
    return c1, c2


def _mutate(ch: str, bit: int) -> str:
    lst = list(ch)
    lst[bit] = '0' if lst[bit] == '1' else '1'
    return ''.join(lst)


def run(seed: int = 13, n_gen: int = N_GEN, p_mut: float = 0.1) -> dict:
    rng_sel = _LCG(seed)
    rng_cx  = _LCG(seed * 7 + 3)
    rng_mut = _LCG(seed * 13 + 5)

    pop = INIT_POP[:]
    best = None
    generations = []

    for g in range(1, n_gen + 1):
        rows = _eval_pop(pop)
        _fitness(rows)

        gen_best = min(rows, key=lambda r: r['f'])
        if best is None or gen_best['f'] < best['f']:
            best = {'chrom': gen_best['chrom'], 'x': gen_best['x'], 'f': gen_best['f'], 'gen': g}

        # selecao por roleta: forma pares de pais
        pairs = []
        children = []
        cross_ops = []
        mut_ops = []
        n_pairs = len(pop) // 2
        for _ in range(n_pairs):
            i1 = _roulette(rows, rng_sel)
            i2 = _roulette(rows, rng_sel)
            p1, p2 = pop[i1], pop[i2]
            pairs.append({'p1': p1, 'i1': i1, 'p2': p2, 'i2': i2})
            if rng_cx.next() <= P_CROSS:
                point = 1 + int(rng_cx.next() * (N_BITS - 1))  # ponto em 1..3
                c1, c2 = _crossover(p1, p2, point)
                cross_ops.append({'p1': p1, 'p2': p2, 'point': point, 'c1': c1, 'c2': c2})
            else:
                c1, c2 = p1, p2
                cross_ops.append({'p1': p1, 'p2': p2, 'point': None, 'c1': c1, 'c2': c2})
            children.extend([c1, c2])

        # mutacao bit a bit
        mutated = []
        for ch in children:
            cur = ch
            for b in range(N_BITS):
                if rng_mut.next() <= p_mut:
                    new = _mutate(cur, b)
                    mut_ops.append({'before': cur, 'bit': b, 'after': new})
                    cur = new
            mutated.append(cur)

        # elitismo: o melhor da geracao atual entra na nova populacao
        new_pop = mutated[:]
        elite = gen_best['chrom']
        new_mut_rows = _eval_pop(new_pop)
        worst_idx = max(range(len(new_mut_rows)), key=lambda i: new_mut_rows[i]['f'])
        replaced = new_pop[worst_idx]
        new_pop[worst_idx] = elite

        generations.append({
            'gen':        g,
            'pop':        pop[:],
            'rows':       rows,
            'gen_best':   {'chrom': gen_best['chrom'], 'x': gen_best['x'], 'f': gen_best['f']},
            'pairs':      pairs,
            'cross':      cross_ops,
            'children':   children[:],
            'mutations':  mut_ops,
            'mutated':    mutated[:],
            'elite':      elite,
            'replaced':   replaced,
            'new_pop':    new_pop[:],
            'best_so_far': dict(best),
        })

        pop = new_pop

    # avaliacao final
    final_rows = _eval_pop(pop)

    return {
        'algorithm': 'genetic_function',
        'config': {
            'expr':   'f(x) = (x-4)^2 - (x-8)^3 + 5',
            'domain': [X_MIN, X_MAX],
            'n_bits': N_BITS,
            'n_gen':  n_gen,
            'p_mut':  p_mut,
            'p_cross': P_CROSS,
            'seed':   seed,
            'objective': 'minimizar',
        },
        'init_pop':    INIT_POP[:],
        'generations': generations,
        'final_pop':   final_rows,
        'best':        best,
    }


if __name__ == '__main__':
    r = run()
    print('f(x) = (x-4)^2 - (x-8)^3 + 5  em [0,15]  (minimizar)')
    print('populacao inicial:', r['init_pop'])
    for g in r['generations']:
        print('\n-- Geracao', g['gen'], '--')
        for row in g['rows']:
            print('  %s  x=%2d  f=%8.2f  fit=%7.2f  p=%.3f' %
                  (row['chrom'], row['x'], row['f'], row['fit'], row['prob']))
        print('  melhor da geracao:', g['gen_best']['chrom'],
              'x=%d f=%.2f' % (g['gen_best']['x'], g['gen_best']['f']))
        print('  nova populacao:', g['new_pop'])
    b = r['best']
    print('\nMelhor solucao: x =', b['x'], '(', b['chrom'], ')  f(x) =', b['f'], ' geracao', b['gen'])
    # conferencia: minimo real por busca exaustiva
    best_x = min(range(16), key=lambda x: _f(x))
    print('Minimo real em [0,15]: x =', best_x, ' f =', _f(best_x))
